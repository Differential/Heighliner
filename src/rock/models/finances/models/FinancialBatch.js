
import uuid from "node-uuid";
import moment from "moment";

import { createGlobalId } from "../../../../util";

import {
  FinancialBatch as FinancialBatchTable,
} from "../tables";

import { Rock } from "../../system";

export default class FinancialBatch extends Rock {
  __type = "FinancialBatch";

  async getFromId(id, globalId) {
    globalId = globalId ? globalId : createGlobalId(id, this.__type);
    return this.cache.get(globalId, () => FinancialBatchTable.findOne({ where: { Id: id }}));
  }

  async findOrCreate({ prefix = "Online Giving", suffix = "", currencyType, date }) {
    let paymentType = "";
    if (currencyType) paymentType = currencyType;

    const batchName = `${prefix} ${paymentType} ${suffix}`.trim();

    const batch = await FinancialBatchTable.find({
      where: {
        Status: 1,
        BatchStartDateTime: { $lte: date },
        BatchEndDateTime: { $gt: date },
        Name: batchName,
      },
    });

    if (batch.length) return batch[0];

    const BatchStartDateTime = moment(date)
      .startOf("day")
      // .subtract(5, "hours")
      .subtract(1, "minute")
      .add(1, "day")
      .toISOString();

    const BatchEndDateTime = moment(date)
      .endOf("day")
      // .subtract(5, "hours")
      .subtract(1, "minute")
      .add(1, "day")
      .toISOString();

    const newBatch = {
      Guid: uuid.v4(),
      Name: batchName,
      Status: 1,
      ControlAmount: 0,
      // XXX this is actually stored on the payment gateway if we want to make
      // it a dynamic value
      BatchStartDateTime,
      BatchEndDateTime,
    };

    const batchId = await FinancialBatchTable.post(newBatch);
    return FinancialBatchTable.findOne({ where: { Id: batchId } });
  }

}
