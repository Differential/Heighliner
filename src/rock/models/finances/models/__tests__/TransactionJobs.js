
import moment from "moment";
import queue from "bull";

import TransactionJobs from "../TransactionJobs";

import {
  Transaction as TransactionTable,
  SavedPayment,
  TransactionDetail,
  ScheduledTransaction,
  ScheduledTransactionDetail,
  FinancialPaymentDetail as FinancialPaymentDetailTable,
  FinancialAccount,
} from "../../tables";

import {
  DefinedValue,
} from "../../../system/tables";

import {
  Person as PersonTable,
  PersonAlias,
} from "../../../people/tables";

import {
  Campus as CampusTable,
  Location as LocationTable,
} from "../../../campuses/tables";

import {
  Group,
  GroupLocation,
  GroupMember,
} from "../../../groups/tables";

import FinancialBatch from "../FinancialBatch";


// import { createGlobalId } from "../../../../../util/node";

jest.mock("../../tables", () => ({
  // FinancialBatch: {
  //   find: jest.fn(),
  //   findOne: jest.fn(),
  //   delete: jest.fn(),
  //   post: jest.fn(),
  // },
}));

jest.mock("../../util/nmi");

jest.mock("moment");

jest.mock("node-uuid", () => ({
  v4: jest.fn(() => "guid"),
}));

jest.mock("bull", () => jest.fn(() => ({
  process: jest.fn(() => {}),
  add: jest.fn(() => {}),
})));

const mockedCache = {
  get: jest.fn((id, lookup) => Promise.resolve().then(lookup)),
  set: jest.fn(() => Promise.resolve().then(() => true)),
  del() {},
  encode: jest.fn((obj, prefix) => `${prefix}${JSON.stringify(obj)}`),
};

it("starts up a transaction queue", () => {
  expect(queue.mock.calls[0][0]).toEqual("Transaction Receipt");
  expect(queue.mock.calls[0][1]).toBe(6379);
});

describe("add", () => {
  let Local;
  beforeEach(() => {
    Local = new TransactionJobs({ cache: mockedCache });
  });
  afterEach(() => {
    Local = undefined;
  });

  it("passes data to the queue", () => {
    Local.queue = {
      process: jest.fn(),
      add: jest.fn(),
    };

    Local.add({ test: "TEST" });
    expect(Local.queue.add).toBeCalledWith({ test: "TEST" }, {
      removeOnComplete: true,
      attempts: 288,
      backoff: { type: "fixed", delay: 60000 * 5 },
    });
  });
});

describe("getOrCreatePerson", () => {
  let Local;
  beforeEach(() => {
    Local = new TransactionJobs({ cache: mockedCache });
  });
  afterEach(() => {
    Local = undefined;
  });

  it("keeps going if a person exists", async () => {
    const Person = {
      Id: 1,
    };

    const data = await Local.getOrCreatePerson({ Person });
    expect(data).toEqual({ Person });
  });
});
