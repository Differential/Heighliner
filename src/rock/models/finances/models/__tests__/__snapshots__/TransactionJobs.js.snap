exports[`batch process correctly orders the functions to be called 1`] = `"function ( /* istanbul ignore next */_ref14) /* istanbul ignore next */{var data = _ref14.data;return (/* istanbul ignore next */_promise2.default.resolve(data).then( /* istanbul ignore next */_this.getOrCreatePerson).then( /* istanbul ignore next */_this.createPaymentDetail).then( /* istanbul ignore next */_this.findOrCreateTransaction).then( /* istanbul ignore next */_this.findOrCreateSchedule).then( /* istanbul ignore next */_this.createTransactionDetails).then( /* istanbul ignore next */_this.createSavedPayment).then( /* istanbul ignore next */_this.updateBillingAddress).then( /* istanbul ignore next */_this.updateBatchControlAmount).then( /* istanbul ignore next */_this.sendGivingEmail));}"`;
